h2o-py/h2o/estimators/estimator_base.py

```diff
@@ -107,14 +108,50 @@ def train(self, x=None, y=None, training_frame=None, offset_column=None, fold_co
         :param float max_runtime_secs: Maximum allowed runtime in seconds for model training. Use 0 to disable.
         :param bool verbose: Print scoring history to stdout. Defaults to False.
         """
-        self._train(x=x, y=y, training_frame=training_frame, offset_column=offset_column, fold_column=fold_column,
-                    weights_column=weights_column, validation_frame=validation_frame, max_runtime_secs=max_runtime_secs, 
-                    ignored_columns=ignored_columns, model_id=model_id, verbose=verbose)
+        parms = self._make_parms(x=x, y=y, training_frame=training_frame, offset_column=offset_column, 
+                                 fold_column=fold_column, weights_column=weights_column, 
+                                 validation_frame=validation_frame, max_runtime_secs=max_runtime_secs, 
+                                 ignored_columns=ignored_columns, model_id=model_id, verbose=verbose)
+        self._train(parms, verbose=verbose)
 
 
-    def _train(self, x=None, y=None, training_frame=None, offset_column=None, fold_column=None,
-              weights_column=None, validation_frame=None, max_runtime_secs=None, ignored_columns=None,
-              model_id=None, verbose=False, extend_parms_fn=None):
+    def bulk_train(self, x=None, y=None, training_frame=None, offset_column=None, fold_column=None,
+                   weights_column=None, validation_frame=None, max_runtime_secs=None, ignored_columns=None,
+                   segments=None, verbose=False):
```

Comment:
ok, from what I understand from `buildModels`, it's actually designed to support combinations, as segments frame can have multiple columns...but still have the feeling that for consistency, it could work in a way similar to `fold_column` as you can always encode multiple columns into one.

Reply:
> is `segments` a kinda standard parameter name for this

I would say in this context segment is commonly used. What people do is sometimes built a model on the full dataset and then "segmented" models to compare performance according to some kind of segmentation (this can be natural or identified by the main model).

Partitioning sounds like a more of a technical term about data representation - would be problematic in the case of Sparkling Water.

> it could work in a way similar to fold_column as you can always encode multiple columns into one

This is true but in reality, (for the users) might be difficult to provide just a single column correctly. It is natural for the segments to be identified by multiple columns.

You make a good point about the analogy with the `fold_column` parameter, however, I would like to stress that this feature is designed in a different way - segments can but don't necessarily exhaust all possible groups (if you grouped by the segment columns) in the input data. You can build on just a subset of the segments. One way to use this is to abort the process and continue with fresh h2o instance only on unfinished segments.

I was also thinking about adding parameter segment_columns which would then work on all the group-bys, but I see this as an addon functionality.

Reply:
> segments can but don't necessarily exhaust all possible groups (if you grouped by the segment columns) in the input data. You can build on just a subset of the segments.

this could also be supported with a single column, where `NA`s stand for "ignore this row", right?

Reply:
> this could also be supported with a single column, where NAs stand for "ignore this row", right?

Yes, it could. At that point, the user can also subset the dataset.

I particularly like this design because it lets me quickly experiment with a single or few models (just make a single row "segments" frame). Do you think this is not a good idea - should this be changed. I like it (obviously :)) but I am might be missing something.

Reply:
There are things that I like in the `segments` frame, mainly the fact that the segmentation semantic is clear: each row defines a segment, couldn't be more clear, and it's a major point.
There are also things that annoy me:
-  the fact that it's also an `H2OFrame` confuses me, as I think it is the first frame that is not used for the training/scoring phase, but I'm open to it, it's just that I don't think we've used frames for parametrization until now.
- as a frame, it needs to repeat some of the column names of the `training_frame`, which I also find a bit confusing: it's not a huge problem per se, just requires more parameter validation.
- I'm not sure how you handle duplicate segments in your approach, as the `unicity` is enforced by the user:
what happens if creates segments like:

| color |
| --- |
| red |
| yellow |
| blue |
| red |
| blue |

or

| color | shape |
| --- | --- |
| red | circle |
| yellow | square |
| red | square |
| red | circle |

do you (want to) handle this on the backend?

one of the advantage of the `fold_column`-like approach, is that you don't have to deal with those issues, the other main being the API consistency which makes it predictable (users will quickly understand how to use it, and have the same expectations as for similar params).

I'm not against the `segments` approach, but I'd rather see it as a sugar on top on client side:
clients call the rapids expr to convert segments (frame or dictionary) into a segment column, adds it to the training_frame and triggers the training.

Reply:
> the fact that it's also an H2OFrame confuses me, as I think it is the first frame that is not used for the training/scoring phase, but I'm open to it, it's just that I don't think we've used frames for parametrization until now.

You are right it is rare. We do use it sometimes (even) for Models, eg. for GLM Beta constraints.

> as a frame, it needs to repeat some of the column names of the training_frame, which I also find a bit confusing: it's not a huge problem per se, just requires more parameter validation.

true, it does require more validation, handling of domains... definitely a disadvantage

> I'm not sure how you handle duplicate segments in your approach, as the unicity is enforced by the user

this can be validated, IMHO not a big deal - worst what can happen is that multiple same models will be built

> one of the advantage of the fold_column-like approach, is that you don't have to deal with those issues, the other main being the API consistency which makes it predictable (users will quickly understand how to use it, and have the same expectations as for similar params).

agreed, but there are also some issues:

- the returned model needs to be identified somehow, it should have information about the segment it was built for - I was planning a 2D table in the Model.Output to identify the segment - with a single generated columns the information is a little bit lost (you could have this column as categorical and the domain can have the encoded multicolumn values into a single domain value) but I don't think we could have the 2d table anymore
- you need to do the same thing for training, validation and any other frame that plays part in the model training (eg. calibration frame), with just a list of segments you don't need to worry

> the fact that it's also an H2OFrame confuses me

we can also allow for other data structures on the input - as long as they can be converted to H2OFrame in the client code and send to the backend as a Frame reference

Reply:
> you need to do the same thing for training, validation and any other frame that plays part in the model training (eg. calibration frame)
yes, for sure the main inconvenient, and probably not a small one given all the places where those columns need to be filtered out, or handled differently (also looking at you `adaptTestFrame`).

> we can also allow for other data structures on the input
was also thinking about a simple `dict` with column names as keys, but even this is a restriction somehow.
My concern here is not only the confusion (training frame vs configration frame) but also because I suspect that most users (Py or R) don't use H2O (a lot) for data munging, and therefore even less to configure estimators. So they're not always familiar with the frame API.
They're more likely to use R frames directly or numpy/pandas, therefore the more native the parameters are, the simpler it is for the users.
So yes, I'm fine as soon as we don't enforce `H2OFrame` type: we should be able to accept anything that is `h2oable` (using `as.h2o` in R or `H2OFrame` constructor in Py).

Reply:
@sebhrusen I added `segment_columns` parameter, this parameter specifies the segment-by columns. I think this is what 99% will be using instead of manually specifying the `segments` Frame.

Are you okay with this approach?

If so I will add documentation and more tests.

Reply:
@michalkurka yes, I like it.
I'd keep one single parameter though and overload it, it's much more common in dynamic languages (pandas and sklearn do this a lot for example) as soon as the functionality of the parameters to be merged is the same. Also, no need to deal with use-cases where user specifies both then...
In your case, I don't think there is any potential confusion between:
```
segments=['sex', 'class'] # use columns directly 
```
and
```
segments=dict(sex=['male', 'female', 'male', 'female'], class=[1, 1, 2, 2])  # define explicit segments
```
for example, so one param would be better.

Reply:
@sebhrusen I did (almost) that, I kept H2OFrame as the alternative and using dict and other structures could be added later. The primary definitely is just to provide the columns to segment-by, everything else is a bonus in this jira/ticket.

Looks okay now?